internal/rbd/rbd_healer.go

```diff
@@ -0,0 +1,184 @@
+		return deviceSecret, err
+	}
+	userID, ok := secret.Data["userID"]
+	if !ok {
+		util.ErrorLogMsg("parse userID failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse userID")
+	}
+	userKey, ok := secret.Data["userKey"]
+	if !ok {
+		util.ErrorLogMsg("parse userKey failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse userKey")
+	}
+	encryptionPassphrase, ok := secret.Data["encryptionPassphrase"]
+	if !ok {
+		util.ErrorLogMsg("parse encryptionPassphrase failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse encryptionPassphrase")
+	}
+	deviceSecret = map[string]string{
+		"encryptionPassphrase": string(encryptionPassphrase),
+		"userID":               string(userID),
+		"userKey":              string(userKey),
+	}
+
+	return deviceSecret, nil
+}
+
+func callNodeStageVolume(ns *NodeServer, c *k8s.Clientset, pv *v1.PersistentVolume, pluginPath string) error {
+	publishContext := make(map[string]string)
+
+	volID := pv.Spec.PersistentVolumeSource.CSI.VolumeHandle
+	if acquired := ns.VolumeLocks.TryAcquire(volID); !acquired {
+		util.ErrorLogMsg(util.VolumeOperationAlreadyExistsFmt, volID)
+		return status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volID)
+	}
+	defer ns.VolumeLocks.Release(volID)
+
+	stagingParentPath := pluginPath + "/kubernetes.io/csi/pv/" +
+		pv.Name + "/globalmount"
+
+	util.DefaultLog("sending nodeStageVolume for volID: %s, stagingPath: %s",
+		volID, stagingParentPath)
+
+	deviceSecret, err := getSecret(c,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Namespace,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Name)
+	if err != nil {
+		util.ErrorLogMsg("getSecret failed for volID: %s, err: %v", volID, err)
+		return err
+	}
+
+	volumeContext := pv.Spec.PersistentVolumeSource.CSI.VolumeAttributes
+	volumeContext["volumeHealerContext"] = "true"
+
+	req := &csi.NodeStageVolumeRequest{
+		VolumeId:          volID,
+		PublishContext:    publishContext,
+		StagingTargetPath: stagingParentPath,
+		VolumeCapability: &csi.VolumeCapability{
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: accessModeStrToInt(string(pv.Spec.AccessModes[0])),
+			},
+		},
+		Secrets:       deviceSecret,
+		VolumeContext: volumeContext,
+	}
+	if pv.Spec.PersistentVolumeSource.CSI.FSType == "block" {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Block{
+			Block: &csi.VolumeCapability_BlockVolume{},
+		}
+	} else {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Mount{
+			Mount: &csi.VolumeCapability_MountVolume{
+				FsType:     pv.Spec.PersistentVolumeSource.CSI.FSType,
+				MountFlags: pv.Spec.MountOptions,
+			},
+		}
+	}
+
+	_, err = ns.NodeStageVolume(context.TODO(), req)
+	if err != nil {
+		util.ErrorLogMsg("nodeStageVolume request failed, volID: %s, stagingPath: %s, err: %v",
+			volID, stagingParentPath, err)
+		return err
+	}
+
+	return nil
+}
+
+// RunVolumeHealer heal the attached volumes.
```

Comment:
This is very Kubernetes specific, it breaks the CSI guidelines as this now will panic on non-Kubernetes container orchestrators. There needs to be a check that logs some error instead of panic'ing (`util.NewK8sClient()`).

Reply:
Agree, is there any easy way to check for CO?

Thanks!

Reply:
@nixpanic Looks like this is not the first place for using `util.NewK8sClient()` it is already used by `genVolFromVolID()`. We should fix it in both places.

I still wonder what would be the best way to identify CO.

Reply:
oh, I thought `util.NewK8sClient()` was only used depending on certain configuration options/features.

You may want to improve `NewK8sClient()` and return an error instead of panic'ing. That would be cleaner in any case, but requires updating all places where is it called.

Reply:
Sounds good Niels, i.e. we should consider improving NewK8sClient() 
I will send that as a separate PR, I have raised a new issue so that we don't miss it later https://github.com/ceph/ceph-csi/issues/2138